File: backend/src/services/daily-content-agent.ts
/**
 * Daily Content Agent – runs on a schedule and fetches today's content automatically.
 * Caches results so the frontend can display them without waiting.
 */
import cron from 'node-cron';
import { getTodayTopic } from '../config/daily-topics.js';

const LITERLY_BACKEND_URL = process.env.LITERLY_BACKEND_URL || 'http://localhost:8000';

export interface CachedTodayContent {
  query: string;
  label: string;
  results: Array<{ title: string; url: string; snippet: string }>;
  fetchedAt: string;
}

let cachedContent: CachedTodayContent | null = null;

async function fetchAndCacheTodayContent(): Promise<void> {
  const topic = getTodayTopic();
  const url = `${LITERLY_BACKEND_URL}/search?q=${encodeURIComponent(topic.query)}&max_results=10`;

  try {
    const response = await fetch(url);
    if (!response.ok) {
      console.error(`[DailyContentAgent] Literly returned ${response.status}`);
      return;
    }

    const data = (await response.json()) as { query: string; results: Array<{ title: string; url: string; snippet: string }> };
    cachedContent = {
      query: data.query,
      label: topic.label,
      results: data.results ?? [],
      fetchedAt: new Date().toISOString(),
    };
    console.log(`[DailyContentAgent] ✅ Fetched and cached: ${topic.label}`);
  } catch (err) {
    console.error('[DailyContentAgent] Fetch failed:', err);
  }
}

export function getCachedTodayContent(): CachedTodayContent | null {
  return cachedContent;
}

export function startDailyContentAgent(): void {
  // Run immediately on startup
  void fetchAndCacheTodayContent();

  // Run every hour (at minute 0)
  cron.schedule('0 * * * *', () => {
    void fetchAndCacheTodayContent();
  });

  console.log('[DailyContentAgent] 🕐 Started – fetches every hour');
}

File: literly-backend/app/services/daily_content_agent.py
"""Daily Content Agent – fetches today's content on schedule and caches it."""
from datetime import datetime

from apscheduler.schedulers.background import BackgroundScheduler

from app.config.daily_topics import get_today_topic
from app.services.perplexity_search import search_perplexity

_cached_content: dict | None = None


def _fetch_and_cache():
    global _cached_content
    topic = get_today_topic()
    try:
        results = search_perplexity(query=topic["query"], max_results=10)
        _cached_content = {
            "query": topic["query"],
            "label": topic["label"],
            "results": results,
            "fetchedAt": datetime.utcnow().isoformat() + "Z",
        }
        print(f"[DailyContentAgent] Fetched and cached: {topic['label']}")
    except Exception as e:
        print(f"[DailyContentAgent] Fetch failed: {e}")


def get_cached_today_content():
    return _cached_content


def start_daily_content_agent():
    _fetch_and_cache()
    scheduler = BackgroundScheduler()
    scheduler.add_job(_fetch_and_cache, "interval", hours=1)
    scheduler.start()
    print("[DailyContentAgent] Started - fetches every hour")
